[Bugfix:Notifications] Grade Inquiry Draft After Submission

hanyunguo opened the pull request:
### Please check if the PR fulfills these requirements:

* [X] Test that for sample- rainbow grades can be configured automatically.
* [X] Test that for sample- rainbow grades can generate grade summaries and customize what to show.
* [X] Check if the results/grades are some for both the instructor and student UI

### What is the current behavior?

Fixes #10247. Currently, when student submit grade inquiries for multiple questions on the same assignment. The text entered in other grade inquiry boxes gets cleared after submitting an inquiry for one question.

### What is the new behavior?

The new behavior introduced by this pull request ensures that the text in the other grade inquiry boxes is not cleared when submitting a grade inquiry for a different question on the same assignment. Users can now submit multiple grade inquiries without losing the entered text, improving the overall user experience and preventing data loss.

### Other information?

The video below demonstrates how to reproduce the issue


https://github.com/Submitty/Submitty/assets/77653905/e53cf416-8dd9-44a5-9f50-450224841c91

and below is new behavior (with saved unsubmitted grade inquiry feature)

https://github.com/Submitty/Submitty/assets/77653905/cc60439d-4734-4b15-896f-cb4c1fb19347

thegame615 reviewed (commented):
Hello,

I was not able to reproduce the issue that this PR is fixing. The issue is that the text disappears when you switch between grade inquiry tabs, correct? I tested it on the most recent main branch and it seems that the text does not get deleted when switched between tabs.  Also, I do not understand your changes to the README.

hanyunguo commented:
> Hello,
> 
> I was not able to reproduce the issue that this PR is fixing. The issue is that the text disappears when you switch between grade inquiry tabs, correct? I tested it on the most recent main branch and it seems that the text does not get deleted when switched between tabs. Also, I do not understand your changes to the README.
Hello, the in main branch does not cause text disappear, it happens when trying to click "Submit Grade Inquiry". For instance, if you have two questions, and you firstly type in both question 1 and question 2. Then, submit only question 1 grade inquiry. The text in question 2 grade inquiry will disappear. Hope it helps. If you still unsure about this issue, I can record a video about this current bug. For the README. changes, I accidentally commit it, I will revert all these changes as soon as possible.

thegame615 commented:
> Hello, the in main branch does not cause text disappear, it happens when trying to click "Submit Grade Inquiry". For instance, if you have two questions, and you firstly type in both question 1 and question 2. Then, submit only question 1 grade inquiry. The text in question 2 grade inquiry will disappear. Hope it helps. If you still unsure about this issue, I can record a video about this current bug. For the README. changes, I accidentally commit it, I will revert all these changes as soon as possible.

If you wouldn't mind, I would like to see a video of the actual bug. I haven't been able to reproduce it yet.

hanyunguo commented:
I've included captions in the video to help explain.

The bug in the most recent main branch version:

https://github.com/Submitty/Submitty/assets/77653905/f712ac56-ed1d-4e62-b815-a869e17f8d08

The version with the bug fixed (PR version):

https://github.com/Submitty/Submitty/assets/77653905/eb6b7c25-120b-430c-918e-526f23616c2c


Also, here's how to produce an assignment with the 'Submit Grade Inquiry' button: 

https://github.com/Submitty/Submitty/assets/77653905/64c057cf-26d8-4cd4-be61-a8ac581a7e6a

thegame615 commented:
> I've included captions in the video to help explain.
> 
> The bug in the most recent main branch version:
> copy_A57B16D3-EF29-4006-89B2-5CAA09C2CF1D.mov
> 
> The version with the bug fixed (PR version):
> New_compressed.mp4
> 
> Also, here's how to produce an assignment with the 'Submit Grade Inquiry' button:
> reproduce_compress.MOV

Thanks for the video.
It's weird, since that is what I tested for and there didn't seem to be a problem on my machine. I'll ask someone else to review it in my stead.

wuj18 reviewed (dismissed):
Hello, I've managed to refactor the issue, and the PR does fix the issue. However, I believe it would be better to remove the console logs.

hanyunguo commented:
Hello,

Thank you for taking the time to review the changes. I've made several improvements:

1. Removed unnecessary console logs to ensure cleaner code.
2. Refactored the code by relocating the function implementation from the <script> section of the Discussion.twig file to the grade-inquiry.js file.
3. Fixed any failed code style check testcases to ensure the code meets quality standards.

Please review the updated code and let me know if there are any further adjustments needed.

04vmatsibekker commented on site/public/js/grade-inquiry.js:
```diff
@@ -1,6 +1,25 @@
+        draftContent = JSON.parse(draftContent);
+    }
+    var elements = document.getElementsByClassName('markdown-textarea fill-available ');
+    for (var i = 0; i < elements.length; i++) {
```
instead of using the variable var, please use let or const instead

hanyunguo replied:
Hello,

I have made some coding style changes to the code. Would you mind taking a moment to review it when you have a chance? Thank you

04vmatsibekker commented on site/public/js/grade-inquiry.js:
```diff
@@ -1,6 +1,25 @@
+        var element = elements[i];
+        var elementId = element.getAttribute('id');
+        const reply_text_area = $(element);
+        if(draftContent.hasOwnProperty(elementId)){
```
fix formatting here

hanyunguo replied:
Thank you for the feedback. I have fixed the formatting.

powe97 commented on site/public/js/grade-inquiry.js:
```diff
@@ -1,6 +1,27 @@
 /* global buildCourseUrl, WebSocketClient */
 /* exported initGradingInquirySocketClient, onComponentTabClicked, onGradeInquirySubmitClicked, onReady, onReplyTextAreaKeyUp */
```
```suggestion
/* exported saveDraft, initGradingInquirySocketClient, onComponentTabClicked, onGradeInquirySubmitClicked, onReady, onReplyTextAreaKeyUp */
```
To fix JS-lint, add `saveDraft` to the exported list.

powe97 commented on site/public/js/grade-inquiry.js:
```diff
@@ -1,6 +1,27 @@
 /* global buildCourseUrl, WebSocketClient */
 /* exported initGradingInquirySocketClient, onComponentTabClicked, onGradeInquirySubmitClicked, onReady, onReplyTextAreaKeyUp */
 
+function saveDraft() { 
```
```suggestion
function saveDraft() {
```
Remove the trailing space here.

powe97 commented on site/public/js/grade-inquiry.js:
```diff
@@ -1,6 +1,27 @@
+        const element = elements[i];
+        const elementId = element.getAttribute('id');
+        const reply_text_area = $(element);
+        if (draftContent.hasOwnProperty(elementId)){
```
```suggestion
        if (Object.prototype.hasOwnProperty.call(draftContent, elementId)) {
```
Change this to use `Object.prototype.hasOwnProperty.call`, because for whatever reason that is what the linter wants.

powe97 commented on site/public/js/grade-inquiry.js:
```diff
@@ -1,6 +1,27 @@
+        const elementId = element.getAttribute('id');
+        const reply_text_area = $(element);
+        if (draftContent.hasOwnProperty(elementId)){
+            var value = draftContent[elementId];
+            reply_text_area.val(value);
```
```suggestion
            reply_text_area.val(draftContent[elementId]);
```
No need for a variable here. Also, use `const` or `let`, never `var`.

powe97 commented on site/public/js/grade-inquiry.js:
```diff
@@ -50,12 +70,25 @@ function onComponentTabClicked(tab) {
+    let draftContent = localStorage.getItem('content');
+    if (draftContent === null) {
+        draftContent = {};
+    } else {
+        draftContent = JSON.parse(draftContent);}
```
```suggestion
    }
    else {
        draftContent = JSON.parse(draftContent);
    }
```

powe97 reviewed (changes requested):
Code looks fine, just need to fix linting issues. I have given suggested changes that should work for fixing this.

williamjallen commented on site/public/js/grade-inquiry.js:
```diff
@@ -30,7 +50,6 @@ function onReady() {
 
 function onComponentTabClicked(tab) {
     const component_id = $(tab).data('component_id');
-
```
This whitespace change should be reverted

hanyunguo replied:
Hi William, I removed the whitespace

williamjallen replied:
It looks like a line is still being deleted here

williamjallen commented on site/public/js/grade-inquiry.js:
```diff
@@ -112,7 +146,6 @@ function onGradeInquirySubmitClicked(button) {
                 const json = JSON.parse(response);
                 if (json['status'] === 'success') {
                     const data = json['data'];
-
```
Same here

williamjallen commented on site/public/js/grade-inquiry.js:
```diff
@@ -1,5 +1,25 @@
+
+function saveDraft() {
+    //saveDraft function displays the saved draft text on grade inquiry box
+    let draftContent = localStorage.getItem('content');
```
`content` isn't a very descriptive key.  Additionally, won't it conflict with other drafts?

hanyunguo replied:
Hi William, I renamed the "content" to "draftContent".

williamjallen commented on site/app/templates/submission/grade_inquiry/Discussion.twig:
```diff
@@ -125,6 +125,9 @@
     <script>
+        window.onload = function() {
+            saveDraft();
+        };      
```
Shouldn't this load the draft instead of saving it?

hanyunguo replied:
Hi William, I changed the function name "saveDraft" to "loadDraft"

williamjallen commented on site/public/js/grade-inquiry.js:
```diff
@@ -1,5 +1,25 @@
+    else {
+        draftContent = JSON.parse(draftContent);
+    }
+    const elements = document.getElementsByClassName('markdown-textarea fill-available ');
```
It would be better to identify elements via a more specific identifier, such as by ID, or by class.  Additionally, it would be better to use jQuery for this.

ZheyuDeng replied:
![image](https://github.com/Submitty/Submitty/assets/122962727/6ece9cd9-8b1b-4e44-b79e-92ecf378d608)
It doesn't seem to have a more specific identifier we can use, I have also switched into jquery.

williamjallen commented on site/public/js/grade-inquiry.js:
```diff
@@ -1,5 +1,25 @@
+        draftContent = JSON.parse(draftContent);
+    }
+    const elements = document.getElementsByClassName('markdown-textarea fill-available ');
+    for (let i = 0; i < elements.length; i++) {
```
You can use [`.forEach()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) for this instead.

williamjallen reviewed (commented):
It looks like this PR is still waiting for some stylistic issues to be addressed.

From the reviewing side, we also need someone to do a functionality review once the stylistic changes are complete.

williamjallen commented on site/app/templates/submission/grade_inquiry/Discussion.twig:
```diff
@@ -125,6 +125,9 @@
     </div>
 
     <script>
+        window.onload = function() {
+            loadDraft();
+        };      
```
Can this be moved to an external JS file?  Also, it would be better to use jQuery for this: `$(document).ready(() => {...`

williamjallen commented on site/public/js/grade-inquiry.js:
```diff
@@ -1,5 +1,23 @@
+
+function loadDraft() {
+    // LoadDraft function displays the saved draft text on grade inquiry box
+    let draftContent = localStorage.getItem('draftContent');
```
This key isn't very unique.  Won't grade inquiries on different gradeables conflict with each other?

williamjallen reviewed (changes requested):
Can you also write a basic Cypress test for this feature?  Small features like this are prone to breaking in non-obvious ways.

zacharymnp reviewed (approved):
works as intended

williamjallen reviewed (approved):
Code seems reasonable to me.
